kliff/dataset/dataset.py

```diff
@@ -394,11 +732,220 @@ def _read(path: Path, weight: Optional[Weight] = None, file_format: str = "xyz")
+                )
+
+            path = to_path(path)
+
+            if path.is_dir():
+                parent = path
+                all_files = []
+                for root, dirs, files in os.walk(parent, followlinks=True):
+                    for f in files:
+                        if f.endswith(extension):
+                            all_files.append(to_path(root).joinpath(f))
+                all_files = sorted(all_files)
+            else:
+                parent = path.parent
+                all_files = [path]
+
+            if len(all_files) == 1:  # single xyz file with multiple configs
+                all_configs = ase.io.read(all_files[0], index=slices)
+                configs = [
+                    Configuration.from_ase_atoms(
+                        config,
+                        weight=copy.copy(weight),
+                        energy_key=energy_key,
+                        forces_key=forces_key,
+                    )
+                    for config in all_configs
+                ]
+            else:
+                configs = [
+                    Configuration.from_ase_atoms(
+                        ase.io.read(f),
+                        weight=copy.copy(weight),
+                        energy_key=energy_key,
+                        forces_key=forces_key,
+                    )
+                    for f in all_files
+                ]
+
+        if len(configs) <= 0:
+            raise DatasetError(
+                f"No dataset file with file format `{file_format}` found at {parent}."
+            )
+
+        logger.info(f"{len(configs)} configurations loaded using ASE.")
         return configs
 
+    def add_from_ase(
+        self,
+        path: Path = None,
+        ase_atoms_list: List[ase.Atoms] = None,
+        weight: Optional[Weight] = None,
+        energy_key: str = "energy",
+        forces_key: str = "forces",
+        slices: str = ":",
+        file_format: str = "xyz",
+    ):
+        """
+        Read configurations from ase.Atoms object and append them to dataset.
+
+        Args:
+            path: Path the directory (or filename) storing the configurations.
+            ase_atoms_list: A list of ase.Atoms objects.
+            weight: an instance that computes the weight of the configuration in the loss
+            energy_key: Name of the field in extxyz/ase.Atoms that stores the energy.
+            forces_key: Name of the field in extxyz/ase.Atoms that stores the forces.
+            slices: Slice of the configurations to read. It is used only when `path` is
+                a file.
+            file_format: Format of the file that stores the configuration, e.g. `xyz`.
+        """
+        configs = self._read_from_ase(
+            ase_atoms_list, path, weight, energy_key, forces_key, slices
+        )
+        self.configs.extend(configs)
+
+    def get_configs(self) -> List[Configuration]:
+        """
+        Get the configurations.
+        """
+        return self.configs
```

Comment:
Make a shallow copy (e.g. `return self.configs[:] `) to avoid modifying the internal data.

Reply:
Done.